controllers/topology/internal/scope/upgradetracker.go

```diff
@@ -0,0 +1,45 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+const maxMachineDeploymentUpgradeConcurrency = 1
+
+// UpgradeTracker is a helper to capture the upgrade status and make upgrade decisions.
+type UpgradeTracker struct {
+	MachineDeployments *MachineDeploymentUpgradeTracker
+}
+
+// NewUpgradeTracker returns a new upgrade tracker.
+func NewUpgradeTracker() *UpgradeTracker {
+	return &UpgradeTracker{
+		MachineDeployments: &MachineDeploymentUpgradeTracker{
+			Names: []string{},
+		},
+	}
+}
+
+// MachineDeploymentUpgradeTracker is a helper to capture the upgrade status and make upgrade
+// decisions for machine deployments.
+type MachineDeploymentUpgradeTracker struct {
+	Names []string
+}
+
+// AllowUpgrade returns true if a machine deployment is allowed to upgrade,
+// returns false otherwise.
+func (m *MachineDeploymentUpgradeTracker) AllowUpgrade() bool {
```

Comment:
Absolutely not sure about this and absolutely feel free to keep it as is:

Would it be better to call this `AllowAdditionalUpgrade`, `AllowNewUpgrade`, or something like this?

I think this could improve expressiveness at the callsite
```go
	if !s.UpgradeTracker.MachineDeployments.AllowUpgrade() {
		return currentVersion, nil
	}
```

The current `AllowUpgrade` sounds a bit like upgrades are entirely blocked (but they are only above a certain threshold).